generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -123,4 +126,39 @@ private static String makeVersionRange(String versionSpec) {
   public static VersionResolver defaultResolver() {
     return new VersionResolver(Aether.defaultOption());
   }
+
+  /**
+   * Takes a version specification (as defined in
+   * http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html) and finds a valid version
+   * that is likely to exist.  Basically: 1.2.3 is 1.2.3+, [1.2.3] is exactly 1.2.3, and then
+   * there is comma-separated range notation.
+   */
+  @Nullable
+  static String resolveVersionLocally(String unparsedVersion) {
+    VersionRange versionRange;
+    try {
+      versionRange = VersionRange.createFromVersionSpec(unparsedVersion);
+    } catch (InvalidVersionSpecificationException e) {
+      return null;
+    }
+    if (versionRange.getRecommendedVersion() != null) {
+      return versionRange.getRecommendedVersion().toString();
+    }
+
+    // There is a range or set of possible versions.
+    for (Restriction restriction : versionRange.getRestrictions()) {
+      // Look for an exact match.
+      if (restriction.getLowerBound().equals(restriction.getUpperBound())) {
+        return restriction.getLowerBound().toString();
+      }
+      // If this is a more complex version restriction, look for an inclusive bound.
+      if (restriction.isUpperBoundInclusive()) {
+        return restriction.getUpperBound().toString();
+      } else if (restriction.isLowerBoundInclusive()) {
+        return restriction.getLowerBound().toString();
+      }
```

Comment:
**Feedback:** Let's locally resolve versions for _only pins_.

The above logic where you look for an inclusive bound introduces two different heuristics for selecting version ranges. One where we look for the highest version, and another where you look for an inclusive bound. _I am not a fan of this._ 

For illustration, suppose `something:something` has versions {1.0, 2.0, 3.0, 3.5, 4.0, 5.0}.  Let's say we make the following requests:
```
something:something:[3.0,)
something:something:[3.0, 5.5),
something:something:(2.9,)
something:something:[3.0, 5.0],
```

These should all evaluate to version 5.0. However, with the above logic, the first two will evaluate to 3.0, and the last two will evaluate to 5.0. 

**Alternative Option**: If you are adamant about resolving version ranges locally, please remove lines 157-158 with the lower bound, and also add a unit tests for the examples above.